### DISCUSSION TITLE
detaching list from session

### DISCUSSION THREAD

**Discussion by the1gofer** [sqlalchemy/sqlalchemy#6867] [Usage Questions]
Is there any way to detach the resulting list of .all() from the session? I'd like to make a funciton that returns a list from the DB, but would rather not have to worry about the session when I do?

I've tried iterating the list and using session.expunge, but that returns None.

**Comment by CaselIT** [comment]
Hi,

Expunge or expunge_all are the right methods.
They removed the original object, so you should not use the return value, but the original objects

**Reply by the1gofer** [reply]
I tried a couple of different things.  First,` return session.expunge(session.query(Source,URL).join(Source.url).filter(URL.processed == True).all())`, which didn't work because  you can't exlunge a list.

Second, I tried

```
        with session_scope_debug() as session:
            sources = session.query(Source,URL).join(Source.url).filter(URL.processed == True).all()
            expunged_list = []
            for source in sources:
                expunged_list.append(session.expunge(source.Source))

            return expunged_list
```

which returns `[None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None]`

So I can't seem to decouple the items and put them in a list either.

**Reply by CaselIT** [reply]
as mentioned, just return the original list, `sources` in your example.

Btw it's probably best to go with mikes suggestion

**Comment by zzzeek** [comment]
why not use a fresh Session each time, then .close() it afterwards ?

**Reply by the1gofer** [reply]
I'm trying to decouple my classes.  

So for example, I have a bunch of url's in the database, and their scraped status.  I thought it would be nice if my scraper class didn't need to know abut and deal with the session.  It could just call something like`URL.unscraped_urls()`, get a list, and do it's thing.

**Reply by zzzeek** [reply]
yes so what im getting at is that you are looking for a "transactionless" flow, where different objects have myobj.get_things(), but there's no notion of "a database transaction" happening throughout.  that is, each "foo.get_things()" method will use a new connection from an engine, run a SELECT, then lose that connection; then you're back in a stateless place until the next "get_something()" operation.    if this is the case then wherever it is that you are creating a query and calling ".all()", that's where you'd also make a new Session object in order to run that query.  then you'd close it, then return your results.   you will get a list of detached objects not connected to any Session.

I know you are talking about "sessions" but the reason "sessions" are so explicit in SQLAlchemy is that we are trying to make the programmer very aware of "connections and transactions".   so just like our query language wants you to be very aware of what SQL you're going for, our session model wants you to be very aware of what "connections / transactions" flow you are going for.    so that's the part you need to define first.   if someone is calling "foo.get_bars()", "bar.get_bats()",  those methods are doing SQL queries, are they all supposed to be in one long running transaction?   or do you want ad-hoc database connections for each one without any state shared?

**Reply by the1gofer** [reply]
Ok.  I don't think I realized I could just return the list after the session is closed.   Thank you!

**Reply by the1gofer** [reply]
actually, I guess may have misunderstood you. 

I thought you meant something like this would work.    @staticmethod
 ```
def processed_sources() -> list:
        """returns list of processed sources"""
        with session_scope_debug() as session:
            sources = session.query(Source, URL).join(Source.url).filter(URL.processed == True).all()

        return sources
```
However, that gives me 

sqlalchemy.orm.exc.DetachedInstanceError: Instance <Source at 0x7fd030048820> is not bound to a Session; attribute refresh operation cannot proceed (Background on this error at: http://sqlalche.me/e/14/bhk3)

reading through that, it seems like I need to do some sort of eager loading.   Do you have any suggestions on which version of eager loading I should use?
